src/js/libgeo/Prover.js

```diff
@@ -68,21 +74,68 @@ function incidentPC(p, c) {
             var erg = General.mult(c.matrix, p.homog);
             erg = General.mult(p.homog, erg);
             erg = CSNumber.abs(erg);
-            if (erg.value.real < 0.0000000000001) {
-                p.incidences.push(c.name);
-                c.incidences.push(p.name);
-            }
+            return (erg.value.real < 0.0000000000001);
         }
     };
 }
 
 function checkConjectures() {
+    var debug = false;
+    if (debug) console.log("conjectures", conjectures.length);
+    //if (!debug)
     if (conjectures.length === 0) return;
-    // TODO: we need some randomized proving here:
-    // move free elements and check conjectures a number of times.
-    // For now assume that all conjectures could be verified.
+    backupGeo();
+
+    var ii, jj, emove;
+    var nummoves = 3;
+
+    // filter free objects which are involved in conjectures
+    var involved = [];
+    conjectures.forEach(function(con) {
+        var invs = con.getInvolved();
+        invs = invs.forEach(function(el) {
+            if (involved.indexOf(el) < 0 && csgeo.free.indexOf(el) >= 0) involved.push(el);
+        });
+    });
+
+    // debug code remove later
+    var nconject = conjectures.length;
+    involved.forEach(function(el) {
+        ii = nummoves;
+        while (ii--) {
```

Comment:
I don't have a good reason for this just now, but my gut feeling would be that it might be better to exchange the order of loops here: perform one move for each of the `involved` elements, and then a second one, and so on up to `nummoves` for each of them. That's what Cinderella does, according to @kortenkamp, and somehow that feels right to me. You *could* consider re-building the list of `involved` elements after each pass, particularly if you managed to disprove some conjecture. Not sure about the performance tradeoff here; I guess it depends on the application.